### PR TITLE
Fix tests

### PR DESCRIPTION
**Pull request by friendlyanon**
Addresses [this comment](https://github.com/David-Haim/concurrencpp/pull/10#issuecomment-701315403).

On the CI pipeline only MSVC compiles and runs the project as of now.  
Please see [this workflow](https://github.com/friendlyanon/concurrencpp/actions/runs/292196457) run for details.

### CODE REVIEW

**Comment by David-Haim** [comment]
Hi there
First of all, thank you for the amount of time and effort you've put into this. 
It will take me a few days to review everything so bare with me :) 

A couple of questions: I see that the CMake file doesn't add any of the example/test/sandbox folders, so they are not built at all. Is this intentional? 
Is this the final PR or do we need another iteration?
Thanks again!

**Comment by friendlyanon** [comment]
Good catch, I forgot about the examples.

Yes, the subprojects are intentionally left out of the root lists file, as they are essentially different projects that consume this library, but are tightly coupled to it.  
If you want everything available from one lists file, you can have an `all` directory in the root with a single lists file that just does `add_subdirectory` for everything you want available from that file.

I would like to say that the reason all the source files are touched as well is because one of the commits in the .patch was to actually run clang-format, but instead of moving the files the .patch seems to have had a delete-create effect.  
If you don't like that I could try to redo that part.

The reason why the commit situation was getting hairy and I made a patch of the changes instead was because of trying to figure out the CI situation, as building across 3 different OSs is pretty tough.

**Comment by David-Haim** [comment]
So a few concrete questions:

1. Do we need install_libcxx.sh, and if so, can it be moved to a ci / cmake folder?
2. How do I compile and launch the tests/sandbox locally? Can it be done without ninja?
3. Can you write a few words about concurrencppConfig.cmake? It's not trivial what it does.
4. there are some TODO's in various files, can you look at them? 
5. The thread sanitizing vs. regular tests: the idea is that we have two kinds of tests: regular and sanitized.
The "regular" tests are compiled and executed normally (without `-fsanitize=thread`). these tests test the functionality, edge-cases and correctness of the library. there are also the thread sanitizer tests, which have to be compiled with `-fsanitize=thread` (along with the library). these tests test some features in isolation or present some exemplary use case that the thread sanitizers can catch data races and such.
So the idea is that the library and tests have to be compiled twice: compile the library + tests normally, compile the library + thread sanitizer tests +  `-fsanitize=thread` flag. run all tests. Currently I think that everything is compiled with `-fsanitize=thread` Am I wrong?

Sorry about the delay, I had busy week.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      fail-fast: false
+      matrix:
+        config:
+          - os: ubuntu-latest
+            cc: clang-10
+            cxx: clang++-10
+            experimental: true
+
+          - os: macos-latest
+            cc: clang
```
Comment (David-Haim) [inline .github/workflows/ci.yml:41]: Is it OK that ubuntu uses clang-10 and macOS doesn't specify a version?

**Review by David-Haim — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (friendlyanon) [reply .github/workflows/ci.yml:41]: [`macos-latest` comes with Clang 10](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md#language-and-runtime) installed, so I don't see a reason to specify a version.

**Comment by friendlyanon** [comment]
<ol><li>From what I understand <code>&lt;experimental/coroutine&gt;</code> lives only in libc++, not libstdc++, so it has to be installed on a *nix system that isn't macOS. Ubuntu here is actually a very inconvenient platform to test, because both libc++ has to be compiled and installed from source and clang-10 has to be installed, so it takes a really long time to run. And that's not counting the fact that the build for *nix platforms will fail anyway, because there seems to be a <a href="https://github.com/friendlyanon/concurrencpp/actions/runs/292196457">codegen issue</a> that doesn't happen in MSVC. Sure, I can move it.</li>
<li>Since the CI pipeline has to work across 3 different platforms, I had to put everything in cmake scripts to make the build process sane. However, the whole process essentially boils down to this:
<pre>cmake -S test -B build/test -D WHATEVER_VARIABLE=...
cmake --build build/test
cd build/test && ctest</pre>
Ninja is optional, it's just faster than make.</li>
<li><code>concurrencppConfig.cmake</code> is installed to <code>&lt;prefix&gt;/&lt;lib&gt;/cmake/concurrencpp-2.0.0/</code>, which is <a href="https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure">one of the paths</a> <code>find_package(concurrencpp)</code> will check for when looking for a config file in config mode. You can put things here like extra functions a dependent might need when working with your library and you must <a href="https://cmake.org/cmake/help/latest/module/CMakeFindDependencyMacro.html"><code>find_dependency</code></a> things you depend on here. If you had a more complex library, you could also provide code here to handle the <code>COMPONENTS</code> argument of the calling <code>find_package</code> command. After all those, you just have to <code>include</code> the file generated by <code>install(EXPORT)</code> command to make your own targets available.</li>
<li>The TODOs I have left in <code>test/CMakeLists.txt</code> are because tests need extensive source level changes to separate them, which wasn't in the scope of my PRs. The two dummy tests are there to still have the old tests run temporarily.</li>
<li>I understand. From the above workflow run you can see that both AppleClang and Clang fail to compile the coroutine code, so TSAN is out of the question until those are fixed. Unfortunately, my knowledge here is very limited, so I don't know what to do about the compilers failing to build.</li>
</ol>

**Comment by David-Haim** [comment]
1. I'm aware of the code-gen bug. luckily it happens only in debug mode. in release mode everything is compiled fine. this is why the tests are still ok to use. 
the thread sanitizer tests have to be compiled in release mode anyway (according to LLVM instructions) because it super-slow otherwise. 

2. this works, at least partially. on Windows, specifying `-DCMAKE_BUILD_TYPE=Release` is ignored. the reason is the MSVC doesn't accept this flag from outside, only from the CMake itself. any workaround?
(Sorry for being annoying, I'm far from being an environment/build specialist so I want to make sure I can still develop this library when you're not around)

5. Same as 1. all of those tests can run in release mode, some of them must be compiled in release mode. for now, no tests in debug mode under clang is something we can live with.

**Comment by friendlyanon** [comment]
Ah, good to know. I had no idea things were only broken in Debug mode.

Since MSVC uses a multi-configuration system, you have to pass the config either via CLI or as an env variable:
```bash
cmake -S . -B build/lib
#
cmake --build build/lib --config Release
# or
CMAKE_CONFIG_TYPE=Release cmake --build build/lib
# for testing the -C flag or the above env variable is needed
ctest -C Release
```
The above is a bash snippet, but I'm sure you can adapt for Windows as well.

**Comment by friendlyanon** [comment]
Okay, [this run](https://github.com/friendlyanon/concurrencpp/runs/1232132713) looks more promising.  
Only AppleClang fails and only at the tests w.r.t building, however the tests don't look so good. Curiously the tests were fine on Ubuntu without TSAN.

**Comment by David-Haim** [comment]
Are these tests run serially? or different test processes run in parallel? (emphasize on test-processes, not hardcoded tests)
The output doesn't make any sense, I see printing of unrelated tests being printed. for example `assertion failed. expected [2] == [6]` cannot be an output of the matrix multiplication TSAN test. It only makes sense if the tests are parallel, and this is a printing of a failed test (because that test expected all the cores to be available, for example). 
If it's not serial, let's re-run it serially once and see the results.

I'm 90% sure this is the problem from looking at the output for a good few minutes :)
```
Start 1: dummy.main
    Start 2: dummy.tsan_executors
    Start 3: dummy.tsan_fibbonacci
1/8 Test actions/toolkit#3: dummy.tsan_fibbonacci ..............   Passed    2.36 sec
    Start 4: dummy.tsan_matrix_multiplication
2/8 Test actions/toolkit#1: dummy.main .........................Child aborted***Exception:   6.17 sec
assertion failed. expected [2] == [6]

    Start 5: dummy.tsan_quick_sort
```

It looks like the main test starts to run, doesn't finish, but then the matrix mlt. test is "stealing" available cores from it, and then tests, which rely on available cores, fail.

After I merge your code, the next step is to refactor the tests to really stabilize them. 

As for the windows output, It also doesn't make any sense, because [make_ready test](https://github.com/friendlyanon/concurrencpp/blob/master/tests/tests/result_tests/make_result_tests.cpp) doesn't have any `assert_equal_smaller` in it. 
Let's serialize the tests and incrementally fix problems.

I have no idea what's wrong in Apple Clang, Maybe its a regression in Clang11? If we degrade it to Clang10/9 maybe it'll be fixed.

**Comment by friendlyanon** [comment]
I enabled the `RUN_SERIAL` property for tests, things seem to work out well now. See [this run](https://github.com/friendlyanon/concurrencpp/actions/runs/299391103).

I looked it up and it seems installing Clang 10 is a lot more involved on macOS, but I'm not sure about that. What I did is to mark that job as experimental, so it doesn't cause the CI to fail.

**Comment by David-Haim** [comment]
I see, it looks really good!
Is there a way to make ctest print stdout and stderr? I noticed that one of the test is a bit shaky and getting to see the output can really help understanding problems in tests.

About macOS, that's saddening, I'll open a bug to clang team soon.
Thanks for the effort and determination.

**Comment by friendlyanon** [comment]
Sure, verbose output can be enabled with the `-V` CLI flag. I'll add it.

**Comment by David-Haim** [comment]
Great, I'll look at everything again and merge this soon.

**Comment by friendlyanon** [comment]
On a second thought, having the macOS configuration in the workflow would mean that the action on the overall Github UI will be concluded as failure, even though on the Actions page the run will have a green checkmark.

It seems there is no way to solve this as of now. [There is an open issue](https://github.com/actions/runner/issues/2347) for this kind of feature, but I guess an actual solution would be to either:
- Remove the macOS configuration
- or figure out how to install Clang 10 on macOS

I have no experience with macOS, so from my perspective the former option would be the easiest.

Edit: maybe the configuration could be removed and instead substituted with a Travis run, as they have a [Xcode 10.3 image](https://docs.travis-ci.com/user/reference/osx/#macos-version) ready to go.

**Comment by David-Haim** [comment]
That's ok.
It will take some time until this will be merged to master, we're merging it to development.
I don't think it's a good idea to add another CI/CD system at this point. 
(what I did with circle ci is experimental and I have to pay in order to use the important stuff, the free plan is not better than github actions..)

Do you need to push any more commits before you can start working on breaking the tests?

**Comment by friendlyanon** [comment]
I haven't looked into Circle CI, but Travis CI is free for public repositories.

The current Github workflow can't stay the way it is, because the macOS configuration will continue to cause CI runs to show up as failed.  
It either has to be removed, fixed by installing Xcode 10.3 or moved to another CI that already comes with that version of Xcode.

Regarding the source files of tests, the goal would be to nicely namespace tests and decompose them to small units.  
I'm not familiar enough with the project to be able to come up with the best grouping, names for groups and decomposition.  
My plan was to lay the groundwork to make use of `ctest` easily.

**Comment by David-Haim** [comment]
It looks like travis ci is actually the best for free tiers.
All the CI/CD vendors only gives 2 cores in the free tier, but travis ci also gives arm architecture for linux.

Are you suggesting splitting the job between github actions and travis ci or move the entire ci/cd to travis?
I am personally more in favor of the latter but I realize it might be too to ask.

**Comment by friendlyanon** [comment]
I don't think there is a reason to stick to just one CI service.  
The jobs for Ubuntu x86_64 and Windows x86_64 are perfectly fine as they are, maybe the Clang 10 and libc++ installation step for Ubuntu could be simplified by creating a prebuilt artifact that just needs to be downloaded and unpacked.

I'll begin working on setting up the Travis workflow.

**Comment by friendlyanon** [comment]
Hurrah! Finally managed to nail it with the Travis CI workflow.  
The library compile fine, there are [issues in the tests](https://travis-ci.org/github/friendlyanon/concurrencpp/builds/734864518#L239) however.

Regarding `arm64` in the CI, I couldn't find prebuilt Clang 10 for `arm64`, so it either has to be built from source (very slow) or you could make a PPA on launchpad with the prebuilt LLVM toolchain to speed things up. The latter could also mean that building libc++ could also be avoided in the Github workflow for ubuntu.

Also, I'd like to warn you that I'm very much not satisfied with the commit history of this PR, so I'd like to delete my fork and redo the changes for a better noise to signal ratio in the commits, now that I figured out this CI business properly.

**Comment by David-Haim** [comment]
I will take a look what's wrong with the compilation, but I'll be a bit rude here and say that the problem is not within the tests but within the compiler :) It's ok with MSVC and Clang on linux but suddenly the same code doesn't compile on apple? 
That's bullshit :laughing: 

If you want to refork the project feel free to do so.

**Comment by friendlyanon** [comment]
Hmm, maybe. With Apple you never know :)

I'll make a new branch and redo the commits there.
